src/UsbPrinterDeviceChannel.ts

```diff
@@ -0,0 +1,134 @@
+import { IPrinterDeviceChannel } from './IPrinterDeviceChannel';
+import { LineBreakTransformer } from './LineBreakTransformer';
+import { PrinterCommunicationMode } from './PrinterCommunicationMode';
+
+/**
+ * Class for managing the WebUSB communication with a printer.
+ */
+export class UsbPrinterDeviceChannel extends EventTarget implements IPrinterDeviceChannel {
+    private device: USBDevice;
+    private deviceIn: USBEndpoint;
+    private deviceOut: USBEndpoint;
+
+    public EnableConsoleDebug = false;
+
+    private commMode: PrinterCommunicationMode;
+    /**
+     * Gets the mode the communication is set up as.
+     */
+    public CommMode() {
+        return this.commMode;
+    }
+
+    private ready: Promise<boolean>;
+    public Ready(): Promise<boolean> {
+        return this.ready;
+    }
+
+    private inputStream: ReadableStream;
+    public StreamFromPrinter() {
+        return this.inputStream;
+    }
+
+    constructor(device: USBDevice, enableConsoleDebug = false) {
+        super();
+
+        this.device = device;
+        this.EnableConsoleDebug = enableConsoleDebug;
+
+        this.ready = new Promise((resolve, reject) => {
+            this.Connect()
+                .then(() => {
+                    resolve(true);
+                })
+                .catch(reject);
+        });
+    }
+
+    public Dispose() {
+        this.device.close();
+    }
+
+    public async SendCommands(commandBuffer: Uint8Array) {
+        if (this.EnableConsoleDebug) {
+            console.debug('Sending print command buffer to printer via USB..');
+            console.debug(commandBuffer);
+            console.time('sendPrintBuffer');
+        }
+
+        try {
+            await this.device.transferOut(this.deviceOut.endpointNumber, commandBuffer);
+        } catch (e: unknown) {
+            const event = new CustomEvent<unknown>('usbPrinterDeviceOutError', e);
+            this.dispatchEvent(event);
```

Comment:
I do not know the complete context around this, but why is this class an `EventTarget`?

Reply:
Based on some googling this [appears to be the right way to implement things that raise events](https://hackwild.com/article/event-handling-techniques/) so other things can add listeners.

Still tweaking the design here a bit but I want to encapsulate the different type of connect/disconnect events that the WebUSB interface can end up generating. So the PrinterDeviceChannel encapsulates the WebUSB specific bits, bubbling them up to the Printer class, which translates them into external events that either the PrinterManager or a custom solution can hook into for displaying popups or something.